### PR TITLE
Add SAM software gpio port

### PR DESCRIPTION
**Pull request by chris-durand**
SAM GPIO improvements:
- [x] Add `SoftwareGpioPort` implementation
- [x] Allow construction of GpioSet from Gpios, not only config classes 
- [x] Add support for inverted pins in GpioSet
- [x] Working hardware unit test for `SoftwareGpioPort`

### CODE REVIEW

**Review by salkinium — APPROVED** [review]
Yes please!

**Comment by salkinium** [comment]
Actually, do you want to update the relevant SAM BSPs with this implementation? I think there [are still some hacks in there](https://github.com/modm-io/modm/blob/develop/src/modm/board/samv71_xplained_ultra/board.hpp#L57).

**Comment by chris-durand** [comment]
> Actually, do you want to update the relevant SAM BSPs with this implementation? 

Done

**Review by salkinium — APPROVED** [review]

File: src/modm/platform/gpio/sam/pin.hpp.in [hunk 1]
```diff
@@ -110,6 +110,25 @@ struct OneOfSignals
...
+
+template<typename T>
+    requires (T::isInverted == true)
+struct is_inverted<T>
+{
+    static constexpr bool value = true;
+};
+
+template<typename T>
+constexpr bool is_inverted_v = is_inverted<T>::value;
```
Comment (salkinium) [inline src/modm/platform/gpio/sam/pin.hpp.in:130]:
```suggestion
constexpr bool is_inverted_v = T::isInverted;
```
I don't fully understand why this helper is needed?

File: src/modm/platform/gpio/sam/pin.hpp.in [hunk 2]
```diff
@@ -143,12 +162,24 @@ class GpioSet : protected PinMuxMixin<PinConfigs...>
...
+	static consteval uint32_t
 	mask(PortName port)
 	{
 		return (((PinConfigs::port == port) ? 1u << PinConfigs::pin : 0u) | ...);
 	}
 
+	static consteval uint32_t
+	invertedMask(PortName port)
+	{
+		return (((PinConfigs::port == port) ? (is_inverted_v<PinConfigs> ? 1u : 0u) << PinConfigs::pin : 0u) | ...);
```
Comment (salkinium) [inline src/modm/platform/gpio/sam/pin.hpp.in:174]:
```suggestion
		return (((PinConfigs::port == port) ? (PinConfigs::isInverted ? 1u : 0u) << PinConfigs::pin : 0u) | ...);
```

File: src/modm/platform/gpio/sam/pin.hpp.in (hunk 1 above)
Comment (chris-durand) [reply src/modm/platform/gpio/sam/pin.hpp.in:130]: What this does in pseudo-code:
```
if T::isInverted is not a valid expression
    return false
else
    return T::isInverted
```
This is to preserve compatibility with passing these gpio config classes to GpioSet for which `T::isInverted` is not a valid expression.

Comment (salkinium) [reply src/modm/platform/gpio/sam/pin.hpp.in:130]: Oh, I see. 👍

**Pull request merged**